rasterio/_err.pyx

```diff
@@ -161,42 +135,61 @@ class GDALError(IntEnum):
     fatal = CE_Fatal
 
 
-cdef class CPLErrors:
-    """A manager for GDAL error handling contexts."""
-
-    def check(self):
-        """Check the error stack and raise or exit as appropriate."""
-        cdef const char *msg_c = NULL
+cdef inline object exc_check():
+    """Checks GDAL error stack for fatal or non-fatal errors
+    
+    Returns
+    -------
+    An Exception, SystemExit, or None
+    """
+    cdef const char *msg_c = NULL
 
-        err_type = CPLGetLastErrorType()
-        # Return True if there's no error.
-        # Debug and warnings are already picked up by the drivers()
-        # context manager.
-        if err_type < 3:
-            CPLErrorReset()
-            return
+    err_type = CPLGetLastErrorType()
+    err_no = CPLGetLastErrorNo()
+    err_msg = CPLGetLastErrorMsg()
 
-        err_no = CPLGetLastErrorNo()
-        msg_c = CPLGetLastErrorMsg()
-        if msg_c == NULL:
-            msg = "No error message."
-        else:
+    if err_msg == NULL:
+        msg = "No error message."
+    else:
         # Reformat messages.
-            msg_b = msg_c
-            msg = msg_b.decode('utf-8')
-            msg = msg.replace("`", "'")
-            msg = msg.replace("\n", " ")
-
-        if err_type == 4:
-            sys.exit("Fatal error: {0}".format((err_type, err_no, msg)))
-        else:
-            CPLErrorReset()
-            raise exception_map.get(err_no, CPLE_BaseError)(
-                err_type, err_no, msg)
-
-    def __enter__(self):
-        CPLErrorReset()
-        return self
+        msg_b = err_msg
+        msg = msg_b.decode('utf-8')
+        msg = msg.replace("`", "'")
+        msg = msg.replace("\n", " ")
 
-    def __exit__(self, exc_type=None, exc_val=None, exc_tb=None):
+    if err_type == 3:
         CPLErrorReset()
+        return exception_map.get(
+            err_no, CPLE_BaseError)(err_type, err_no, msg)
+
+    if err_type == 4:
+        return SystemExit("Fatal error: {0}".format((err_type, err_no, msg)))
+
+    else:
+        return
+
+
+cdef int exc_wrap_int(int err) except -1:
+    """Wrap a GDAL/OGR function that returns CPLErr or OGRErr (int)
+
+    Raises a Rasterio exception if a non-fatal error has be set.
+    """
+    if err:
+        exc = exc_check()
+        if exc:
+            raise exc
+            return -1
```

Comment:
Yep, it's the way to signal exceptions with `cdef` functions that return anything other than `object`: http://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values.